Work around Metal struct padding

shoaibkamil opened the pull request:
Metal introduces padding in structs to preserve alignment, even if the struct is ordered by size as ours is.  The lack of (apparently undocumented) padding in our code triggers asserts in Apple's runtime if Metal Validation is enabled.  To work around this, we pad to a multiple of the largest component of the struct.

@zvookin PTAL.

abadams commented:
I think those bugs were fixed in master. Pushed a merge.

abadams commented on src/runtime/metal.cpp:
```diff
@@ -700,10 +701,26 @@ WEAK int halide_metal_run(void *user_context,
+        //
+        // TODO(shoaibkamil): Follow up with Apple to find a specification for when
+        // and how much padding is needed
+        size_t padded_args_size = (total_args_size + arg_sizes[0] - 1) & -arg_sizes[0];
```
Is & -arg_sizes[0] a common idiom? I had to think way too hard about that. I think elsewhere in the compiler we do & ~(arg_sizes[0] - 1)

shoaibkamil commented:
Gentle review ping; the failure is due to OpenCL issues on the Windows buildbot.

zvookin commented on src/runtime/metal.cpp:
```diff
@@ -717,10 +718,17 @@ WEAK int halide_metal_run(void *user_context,
-        if (total_args_size < 4096 && metal_api_supports_set_bytes) {
+        // The Metal compiler introduces padding up to a multiple of 4 bytes
+        // in the struct, per email communication from Apple
+        size_t padded_args_size = (total_args_size + 4 - 1) & ~(4 - 1);
```
I'd like to change this to "... & ~((size_t)(4 - 1));" (What happens currently on systems with 32-bit ints and 64-bit pointers is the expression produces 0xfffffffc and that gets sign extended to the 64-bit value, but this depends on twos complement, etc. Better to make both the use of unsigned and "same size as a pointer" clear.)

shoaibkamil replied:
Good catch-- fixed, now.

zvookin commented:
The only change here is to round up the total size to a multiple of four bytes right? LGTM pending passing tests and the one comment I made.

shoaibkamil commented:
@zvookin Yes, that's the only change here.  I've pushed a commit addressing your comment as well.

I think @steven-johnson has pulled in the buildbot changes as well, enabling Metal Validation.

zvookin commented:
Looks like there is an actual Metal failure due to validation.(see mac-64-testbranch-401 .) I don't think it has to do with this change, but I'd rather not merge this until we figure out the failure and have tests pass cleanly.

Also, is the error this is fixing already covered by a test? I believe so, but wanted to make sure.

zvookin commented:
I think this reduces the number of Metal validation failures. We'll have to catch the others that are extant in soon-to-come PRs.
